include/unifex/find_if.hpp

```diff
@@ -24,63 +26,219 @@
+      // alive for the duration.
+      // Once we implement it, replace this with let_with which would allocate data
+      // into the operation state directly to avoid the heap allocation.
+      // Use a two phase process largely to demonstrate a simple multi-phase algorithm
+      // and to avoid using a cmpexch loop on an intermediate iterator.
+      return unifex::let(
+        unifex::just(std::vector<Iterator>(num_chunks, end_it), std::forward<Values>(values)...),
+        [func = std::move(func_), sched = unifex::get_scheduler(receiver), begin_it,
+         chunk_size, end_it, num_chunks, found_flag = std::move(found)](
+            std::vector<Iterator>& perChunkState, Values&... values) mutable {
+          // Inject a stop source and make it available for inner operations.
+          // This stop source propagates into the algorithm through the receiver,
+          // such that it will cancel the bulk_schedule operation.
+          // It is also triggered if the downstream stop source is triggered.
+          return unifex::let_with_stop_source([&](unifex::inplace_stop_source& stopSource) mutable {
+            auto bulk_phase = unifex::bulk_join(
+                unifex::bulk_transform(
+                  unifex::bulk_schedule(std::move(sched), num_chunks),
+                  [&](std::size_t index){
+                    auto chunk_begin_it = begin_it + (chunk_size*index);
+                    auto chunk_end_it = chunk_begin_it;
+                    if(index < (num_chunks-1)) {
+                      std::advance(chunk_end_it, chunk_size);
+                    } else {
+                      chunk_end_it = end_it;
+                    }
+
+                    for(auto it = chunk_begin_it; it != chunk_end_it; ++it) {
+                      if(std::invoke(func, *it, values...)) {
+                        // On success, store the value in the output array
+                        // and cancel future work.
+                        // This works on the assumption that bulk_schedule will launch
+                        // tasks (or at least, test for cancellation) in
+                        // iteration-space order, and hence only cancel future work,
+                        // to maintain the find-first property.
+                        perChunkState[index] = it;
+                        *found_flag = true;
+                        stopSource.request_stop();
+                        return;
+                      }
+                    }
+                  },
+                  unifex::par
+                )
+              );
+            return
+              unifex::transform(
+                unifex::transform_done(
+                  std::move(bulk_phase),
+                  [&stopSource, &found_flag](){
+                    if(*found_flag == true) {
+                      // If the item was found, then continue as if not cancelled
+                      return just();
```

Comment:
How can this return both just() and just_done()?

Edit: Ah, I see what you did there... :s

Reply:
Yes... that's a fair complaint :) I think there is value in just not needing the magic to have multiple variants though.